clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -477,6 +475,7 @@ private long waitOnMetadata(String topic, long maxWaitMs) throws InterruptedExce
         long remainingWaitMs = maxWaitMs;
         while (metadata.fetch().partitionsForTopic(topic) == null) {
             log.trace("Requesting metadata update for topic {}.", topic);
+            this.metadata.add(topic);
```

Comment:
The additional call is needed since topics are now removed from the metadata when UNKNOWN_TOPIC_OR_PARTITION response is received. The topic is added again in the retry loop to ensure that it is included in the metadata topic list. Have added a comment.